### PR TITLE
sdk: deduplicate sending read receipts

### PR DESCRIPTION
**Pull request by bnjbvr**
While looking at a rageshake from EX apps, I've noticed that we can now get multiple requests to send the same read receipt (which is expected since I've asked EX devs to completely let us handle read receipts, to avoid stuck unreads). It's wasted work, and we already have a deduplicating handler somewhere in the code (which could be a separate crate useful in isolation, maybe?).

### CODE REVIEW

File: crates/matrix-sdk-ui/src/timeline/mod.rs [hunk 1]
```diff
@@ -718,8 +723,19 @@ impl Timeline {
             return Ok(false);
         }
 
-        trace!("sending receipt");
-        self.room().send_single_receipt(receipt_type, thread, event_id).await?;
+        // Since neither `ReceiptType` nor `ReceiptThread` implement Hash/Ord, we
+        // "flatten" them to a string, which we use to uniquely identify the
+        // query.
+        let request_key = format!("{}|{}", receipt_type, thread.as_str().unwrap_or("<unthreaded>"));
```
Comment (poljar) [inline crates/matrix-sdk-ui/src/timeline/mod.rs:729]: Hmm, why not put the deduplication into the `send_single_receipt()` method?

Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/mod.rs:729]: ... Very good point :grin:

Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/mod.rs]: That was a great idea, and I don't even need to make the deduplicating handler public now. Thanks!

**Review by poljar — APPROVED** [review]
Looks good, not sure if we have a test for this, would be nice to add if not, but not required.

*edit: Oh why did you remove the docs on the deduplicate handlers, they were useful.

**Comment by bnjbvr** [comment]
> Looks good, not sure if we have a test for this, would be nice to add if not, but not required.

Yeah, this is indirectly tested by the timeline code, and other code making use of `send_read_receipt`. Also we have orthogonal tests for the deduplicating handler, so I won't add tests that specifically check if we do deduplicate.

> *edit: Oh why did you remove the docs on the deduplicate handlers, they were useful.

Thanks, today's the first time I've used reflog to resurrect a dropped commit :)

**Comment by poljar** [comment]
> > *edit: Oh why did you remove the docs on the deduplicate handlers, they were useful.
> 
> Thanks, today's the first time I've used reflog to resurrect a dropped commit :)

Congrats, albeit confusing, it is a useful tool.

**Pull request merged**